sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/geneformer_mlm_loss_eval.py

```diff
@@ -0,0 +1,295 @@
+
+
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: LicenseRef-NvidiaProprietary
+#
+# NVIDIA CORPORATION, its affiliates and licensors retain all intellectual
+# property and proprietary rights in and to this material, related
+# documentation and any modifications thereto. Any use, reproduction,
+# disclosure or distribution of this material and related documentation
+# without an express license agreement from NVIDIA CORPORATION or
+# its affiliates is strictly prohibited.
+
+import argparse
+import functools
+import pickle
+from copy import deepcopy
+from pathlib import Path
+from typing import Dict, Type
+
+import torch
+import torch.distributed
+import torch.utils
+import torch.utils.data
+from megatron.core.transformer.module import Float16Module
+from nemo.utils import logging
+from torch.utils.data import DataLoader
+from tqdm import trange
+from transformers import AutoModelForMaskedLM
+
+from bionemo.core.data.multi_epoch_dataset import EpochIndex
+from bionemo.core.utils.dtypes import get_autocast_dtype
+from bionemo.geneformer.api import GeneformerConfig
+from bionemo.geneformer.data.singlecell.dataset import SingleCellDataset
+from bionemo.geneformer.data.singlecell.preprocess import GeneformerPreprocess
+from bionemo.geneformer.tokenizer.gene_tokenizer import GeneTokenizer
+from bionemo.llm.data import collate
+from bionemo.llm.model.biobert.model import BioBertConfig
+from bionemo.testing import megatron_parallel_state_utils
+from bionemo.testing.data.load import load
+
+
+class GeneformerHFAdapter(torch.nn.Module):
+    """An adapter class for running the HF model against our subset of tokens."""
+
+    def __init__(self, hf_path: str, my_token_dict: Dict[str, int], nv_tokenizer: GeneTokenizer):
+        """An adapter that filters and re-orders tokens to match our tokenizer but with the original indices."""
+        super().__init__()
+        self.model = AutoModelForMaskedLM.from_pretrained(hf_path)
+        self.my_token_dict = deepcopy(my_token_dict)
+        self.nv_tokenizer = deepcopy(nv_tokenizer)
+        self.n_tokens_nv = len(self.nv_tokenizer.vocab)
+        self.n_tokens_hf = len(my_token_dict)
+
+        # nvidia tokenizer has [cls] and [pad] first along with some others that do not overlap. This mapper
+        hf_ordered_nv_tokenizer = {
+            self.nv_tokenizer.pad_token: my_token_dict["<pad>"],
+            self.nv_tokenizer.mask_token: my_token_dict["<mask>"],
+            self.nv_tokenizer.cls_token: my_token_dict["<cls>"],
+            self.nv_tokenizer.sep_token: my_token_dict["<eos>"],  # name doesn't really matter here
+        }
+        tokens = list(my_token_dict.items())
+        for k, t in tokens[:4]:
+            assert k.startswith("<")
+
+        missing_nv_tokens = []
+        extra_tokens_not_covered = []
+        for ens, idx in list(my_token_dict.items())[4:]:
+            assert ens.startswith("ENSG")
```

Comment:
Where is ENSG coming from?

Reply:
This is a sanity check that the user is pointing their data at ENSG (ensembl) format gene ids.

Reply:
I misunderstood from a comment lower down.